compiler/rustc_passes/src/intrinsicck.rs

```diff
@@ -369,16 +369,29 @@ impl<'tcx> ExprVisitor<'tcx> {
             // Note that this is only possible for explicit register
             // operands, which cannot be used in the asm string.
             if let Some(reg) = op.reg() {
+                // Some explicit registers cannot be used depending on the
+                // target. Reject those here.
+                if let InlineAsmRegOrRegClass::Reg(reg) = reg {
+                    if let Err(msg) = reg.validate(
+                        asm_arch,
+                        self.tcx.sess.relocation_model(),
```

Comment:
Would it make sense to somehow make register reservations be part of the `asm_target_features` calculation? As both are (always?) used together, might as well return a structure that bundles the target features with the relocation model and maybe some other frequently shared information (target?) too.

Reply:
I tried implementing this but I'm having some trouble working with the query system. I added an `asm_target_info` query which returns this struct:
```rust
#[derive(Debug, Clone, HashStable_Generic)]
pub struct InlineAsmTargetInfo<'a> {
    pub reloc_model: RelocModel,
    pub target_features: FxHashSet<Symbol>,
    pub target: &'a Target,
}
```

Specifically, I can't manage to get the `HashStable` derive work with this struct. Perhaps I'm using the query system wrong?

Reply:
Hm, okay, lets leave it as is for the time being.

Having thought about it further, I'm no longer sure duplicating `Target` and `RelocModel` in the query system outputs for a specific `DefId` is the correct model, anyway.